reduce.1
.TH reduce local 2015-10-22 local  "local doc"
.hy 0 
.if n .ad l 
.SH NAME
reduce \- clean and filter a multiple sequence alignment
.SH SYNOPSIS
.nf
.B reduce \fB[\fP\fB-sv\fP\fB][\fB\-a \fI\sacred_file\fR ] in.msa in_distance_matrix.hat2 out.msa n_to_keep
.SH DESCRIPTION
We have a huge multiple sequence alignment that is too big for building a good tree. This program will read the distance matrix from mafft, visit pairs of sequences that are most similar and remove one member of each pair. If there are sequences which are especially interesting, you can declare them to be sacred. If they appear within a pair, they will not be removed.
.PP
From each pair, we can also choose the strategy for removing members. By default, we keep the first sequence from every pair, but this can be changed with the
.B \-c
option.
.PP
The program keeps removing sequences until there are no more than
.I n_to_keep
left.
.PP
Be default, completely empty columns will be removed (see
.BR \-f " flag".
.SH OPTIONS
.TP 7
\fB-a\fP \fIfilename\fP
\fIfilename\fP is a list of sequences which are sacred and should not be deleted. Sequences should be listed one per line. They should appear exactly as in the
.I in.msa
file, with the comment character like this
.nf
> AAX51857.1 putative profilin [Toxoplasma gondii]
> NP_001015592.1 profilin-1 [Bos taurus]
.fi

.TP 7
.BI \-c " method"
From pairs of sequences, we have to pick which to delete. This is done according to
. Which can be one of
.RS
.IP first 14
From each pair, choose the first.
.IP second 14
From each pair, choose the second.
.IP random 14
From each pair, choose one pseudo-randomly.
.RE
.TP 7
.BI \-e " seed"
If using the random choice method for picking sequences to keep, set the random seed with this option. If you do not do so, there will be a default value so you will get reproducibility from run to run.
.TP
.B \-f
If you want, you can keep completely empty columns in an alignment. It is hard to see why this would be useful except for debugging.
.TP
.B \-g
Remove gaps from sequences when they are written out. Use this option if you want to re-align the sequences after reducing the set.
.TP
.B \-i
Do not check the length of sequences as they are read up. Ignore the sequence length. This is for the unlikely case that you have distances, presumably calculated by a multiple sequence alignment, but you want to do the reduction on a file of unaligned sequences.
.br
This option is not compatible with the
.B  \-f
option.
.TP
.BI \-p " plotfilename"
Every time, before a sequence is removed, print out the number of sequences remaining and the corresponding distance from the distance matrix. You can then plot them. Output goes to
.IR plotfilename .
If the filename is \fI-\fR, output will go to standard output. To plot the data in a program like \fIgnuplot\fR, you might say,
.in +4n
.EX
plot plotfilename using 1:2
.EE
.in
but you probably want to reverse the xrange, so it goes from biggest to smallest.
.in +4n
.EX
set xrange [] reverse
.EE
.in
.TP 7
\fB-s\fP
Sequences containing "seed" will be removed. Mafft uses these as constraints on the alignment. They usually come from structural alignments.
.TP 7
\fB-v\fP
Be more verbose. Multiple options increase verbosity.
.SH NOTES
We use mafft for multiple sequence alignments, so we start by reading their formats.
The biggest issue is the ordering of sequences.
Mafft writes the distance matrix in the same order as its input sequences. It may then reorder the sequences when it writes the alignment, which is what we read. This means that the order in the distance matrix file may not correspond to the order in the alignment.

This is not fatal. It means we have to parse the names written by mafft at the start and use these as lookups.
.SS Threads
There are a few threads here. Input files are read concurrently. This is a good idea. The first time sequences are read, they are gobbled up and put in a queue where they are processed, gaps counted and put into a map object. This was fun, but may not save much time.
.SS Storage
Sequences are read up and indexed by their comment field in a map structure. This was not such a good idea. It might have been better to use a simple array and helper arrays to allow us to handle files with re-ordered sequences.

.SH EXAMPLES
You have calculated an alignment and put it in
.IR "aligned_seqs.fa" .
You do not care which sequences are kept within each pair, but you have two sequences which must be kept and you want to see them in later work. In a text editor, you make a file called
.I sacred_seqs
and it has a few lines like
.nf
> AAX51857.1 putative profilin [Toxoplasma gondii]
> NP_001015592.1 profilin-1 [Bos taurus]
.fi
From an alignment with
.I mafft
, you have a distance matrix file with a name like
.IR "aligned_seq.fa.hat2" .
run
.in +4n
.EX
reduce -a sacred_seqs aligned_seqs.fa aligned_seqs.fa.hat2 reduced_seqs.fa 1000
.EE
.in
and you will get 1000 sequences in
.IR reduced_seqs.fa .

Now, imagine you do not know how many sequences you should keep. You see how the size of matrix entries goes up as you remove sequences. Then you might say
.in +4n
.EX
reduce -p distplot.dat aligned_seqs.fa aligned_seqs.fa.hat2 /dev/null 1
.EE
.in
You will get a file for plotting in \fIdistplot.dat\fR, and you will not get any sequences written out (no reduced set), because the output is put in /dev/null.
.SH BUGS
.IP * 
To get items off the queue, we do a
.nf
  front()
  pop()
.fi
They are combined into front_and_pop(), but there is some unnecessary copying.

.IP *
If a program like jalview is allowed to work with the sequences, he may well take a line like
.br
.nf
> abcde
.fi
and rewrite it like
.br
.nf
>abcde
.fi
This code will expect the space in the sequence file and complain about a mismatch with the distance file. Workaround: Do not store full comments. Store stuff after the leading ">" and after discarding white space. This may break other programs that expect the ">" to be present.

.IP *
There should be another option for choosing the member of each pair to be kept. We should offer to pickt the longer or shorter. The choice is not obvious. Often the longer just has rubbish such as unrecognised introns, so we want the shorter. Unfortunately, the shorter might be a partial sequence.


reduce.cc
/* 8 oct 2015
 * I have an MSA and want to
 *  - get rid of sequences which seem to have monster insertions
 *  - get rid of empty columns
 * Overall scheme
 *   Read the MSA, but only store the comments and put them in a hash.
 *   Read a list of magic sequences, that must be kept.
 *   Read the distance matrix and build a list.
 *   Go back to the MSA, copy entries to keep in to the output file.
 * TODO
 * - very important - when we are given a bad sequence, the whole thing crashes !
 *   shut the queue down on error.
 * - add a check when reading sequences for HHHHH. If one is found, print out a warning.
 * - add an option to remove all gap characters and rewrite the sequences
 */

#include <cerrno>
#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <unistd.h>
#include <random>
#include <queue>
#include <map>
#include <vector>

#include "bust.hh"
#include "distmat_rd.hh"
#include "fseq.hh"
#include "fseq_prop.hh"
#include "mgetline.hh"
#include "plot_dist_reduce.hh"
#include "t_queue.hh"

using namespace std;

/* ---------------- structures and constants ----------------- */
static const char GAPCHAR = '-';
static const unsigned N_SEQBUF = 500;
static const unsigned SEQ_LINE_LEN = 60; /* How many chars per line output seqs */
static const char    *SEED_STR = "_seed_"; /* mafft marker for seed alignments */

static const unsigned char NOBODY = 0;
static const unsigned char S_1    = 1;
static const unsigned char S_2    = 2;

static const int DFLT_SEED = 180077;

struct seq_props {
    map<string, fseq_prop> f_map;
    size_t len;
};

/* ---------------- usage ------------------------------------ */
static int usage ( const char *progname, const char *s)
{
    static const char *u
        = ": [-fgsv -a sacred_file -c choice -e seed] mult_seq_align.msa \
dist_mat.hat2 outfile.msa n_to_keep";
    return (bust(progname, s, "\n", progname, u, 0));
}

/* ---------------- from_queue -------------------------------
 * We have bundles of sequences in a vector. Pull each from
 * the queue
 */
static void
from_queue (t_queue <fseq> &q_fs, map<string, fseq_prop> &f_map)  {
    unsigned n = 0;
    while (q_fs.alive()) {
        fseq fs = q_fs.front_and_pop();
        fseq_prop f_p (fs);
        f_map [fs.get_cmmt()] = f_p;
        n++;
    }
    cout << __func__ << " read "<< n<< " seqs\n";
}

/* ---------------- get_seq_list -----------------------------
 * From a multiple sequence alignment, read the sequences and
 * just fill out the information.
 * If we are going to put this in a thread, we have to catch exceptions
 * here
 */
static void
get_seq_list (struct seq_props & s_props, const char *in_fname,
              const bool ignore_len_check, int *ret) {
    string errmsg = __func__;
    size_t len_check;
    ifstream infile (in_fname);
    *ret = EXIT_SUCCESS;
    if (!infile) {
        *ret = EXIT_FAILURE;
        return (bust_void(__func__, "opening", in_fname, strerror(errno), 0));
    }
    { /* look at the first sequence, get the length, so we can check on */
        streampos pos = infile.tellg();             /* subsequent reads */
        fseq fs(infile, 0);
        len_check = fs.get_seq().size();
        infile.seekg (pos);
    }
    s_props.len = len_check;
    t_queue <fseq> q_fs(N_SEQBUF);
    thread t1 (from_queue, ref(q_fs), ref(s_props.f_map));

    {
        fseq fs;
        unsigned scount = 0;
        if (ignore_len_check)
            len_check = 0;
        try {
            for (; fs.fill(infile, len_check); scount++)
                q_fs.push (fs);
        } catch (runtime_error &e) {
            cerr<< "problem reading sequences\n"<< e.what()<<"\n";
            *ret = EXIT_FAILURE;
            goto clean_and_stop;
        }
        cout << "read "<< scount << " seqs\n";
    }
clean_and_stop:
    infile.close(); /* Stroustrup would just wait until it went out of scope */
    q_fs.close();
    t1.join();
}

/* ---------------- get_sacred -------------------------------
 * read the file of sacred sequences.
 * Could be called as a thread, so no exceptions allowed and
 * put success/failure in the last argument.
 */
static void
get_sacred (const char *sacred_fname, vector<string> &v_sacred, int *sacred_ret)
{
    ifstream sac_file (sacred_fname);
    string s;
    if (!sac_file) {
        *sacred_ret = EXIT_FAILURE;
        return(bust_void(__func__, "opening", sacred_fname, ":", strerror(errno), 0));
    }
    while (mgetline (sac_file, s))
        v_sacred.push_back (s);
    sac_file.close();
    v_sacred.shrink_to_fit();
    *sacred_ret = EXIT_SUCCESS;
}

/* ---------------- check_lists ------------------------------
 * Make sure that the sequences given in the distance file
 * exist in the alignment file, as store in f_map.
 */
static int
check_lists ( const map<string, fseq_prop> &f_map, vector<string> &v_cmt)
{
    unsigned n = 1;
    const char *s1 = "\" in distmat file not found\nIt was sequence number ";
    for (vector<string>::const_iterator it = v_cmt.begin(); it != v_cmt.end(); it++, n++)
        if (f_map.find(*it) == f_map.end())
            return(bust(__func__, "Sequence \"", it->c_str(), s1, to_string(n).c_str(), 0));
    return EXIT_SUCCESS;
}

/* ---------------- mark_sacred ------------------------------
 * Take each sequence that has been marked as sacred.
 * Look for it in the f_map. If we find it, mark it as sacred.
 */
static int
mark_sacred (map<string, fseq_prop> &f_map, vector<string> &v_sacred)
{
    vector<string>::const_iterator it = v_sacred.begin() ;
    int ret = EXIT_SUCCESS;
    for ( ;it != v_sacred.end(); ++it) {
        if (f_map.find(*it) == f_map.end())
            return (bust(__func__, "seq starting", it->c_str(), "not found", 0));
        f_map[*it].make_sacred();
    }
    return ret;
}

/* ---------------- decider functions ------------------------
 * We have several possibilities for deciding which member of a
 * pair to delete. We use one of these, as pointed to by a
 * function pointer.
 */
typedef unsigned char decider_f (const fseq_prop&, const fseq_prop&, default_random_engine&);
#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wunused-parameter"
static unsigned char
always_first (const fseq_prop &f1, const fseq_prop &f2, default_random_engine &r_engine)
{
    return S_1;
}

static unsigned char
always_second (const fseq_prop &f1, const fseq_prop &f2, default_random_engine &r_engine)
{
    return S_2;
}

static unsigned char
decide_random(const fseq_prop &f1, const fseq_prop &f2, default_random_engine &r_engine)
{
    uniform_int_distribution<int> d{0,1};
    int r = d(r_engine);
    if (r == 1)
        return S_1;
    return S_2;
}

static unsigned char
decide_longer(const fseq_prop &f1, const fseq_prop &f2, default_random_engine &r_engine)
{
    if (f1.ngap < f2.ngap)
        return S_2;
    return S_1;
}
#pragma GCC diagnostic pop
static decider_f *
set_up_choice (const string &s)
{
    map <const string, decider_f *> choice_map;
    choice_map["first"]  = always_first;  /* Table of keywords and */
    choice_map["second"] = always_second; /* function pointers */
    choice_map["random"] = decide_random;
    choice_map["longer"] = decide_longer;
    const map<const string, decider_f *>::const_iterator missing = choice_map.end();
    const map<const string, decider_f *>::const_iterator ent = choice_map.find(s);

    if (ent == missing) {
        cerr << __func__ <<": random choice type \"" << s
             << "\" not known. Stopping. Try one of..\n   ";
        map<const string, decider_f *>::const_iterator it = choice_map.begin();
        for (; it != missing; it++)
            cout << it->first << " " << '\n';
        cout << endl;
        return nullptr;
    }
    return ent->second;
}

/* ---------------- choose_seq   -----------------------------
 * Given two seq props, choose one for deletion. Return 1 or 2
 * or 0 for nobody.
 * We will have to expand this to consider different options.
 */
static unsigned char
choose_seq (const fseq_prop &f1, const fseq_prop &f2, decider_f *choice, default_random_engine &r_engine)
{
    if (f1.is_sacred() && f2.is_sacred())
        return NOBODY;
    if (f1.is_sacred())
        return S_2;
    else if(f2.is_sacred())
        return S_1;
    return (choice (f1, f2, r_engine));
}

/* ---------------- remove_seq -------------------------------
 * Walk down the list of distances (v_dist), deciding who to delete.
 * We get the indices of the two sequence in ndx1 and ndx2. We look for these
 * in f_map.
 */
static void
remove_seq (map<string, fseq_prop> &f_map, vector<string> &v_cmt,
            vector<dist_entry> &v_dist, const unsigned long to_keep,
            decider_f *choice, default_random_engine &r_engine)
{
    vector<dist_entry>::const_iterator it = v_dist.begin();
    for ( ; f_map.size() > to_keep  && (it != v_dist.end()); it++) {
        const string &s1 = v_cmt[it->ndx1];
        const string &s2 = v_cmt[it->ndx2];
        const map<string, fseq_prop>::const_iterator missing = f_map.end();
        const map<string, fseq_prop>::const_iterator f1      = f_map.find(s1);
        const map<string, fseq_prop>::const_iterator f2      = f_map.find(s2);
        if ((f1 == missing) || (f2 == missing)) /* sequence already removed */
            continue;                         /* from the list of sequences */
        switch (choose_seq(f1->second, f2->second, choice, r_engine)) {
        case NOBODY:
            continue;        /* break; otherwise compiler complains */
        case S_1:
            distplot (f_map.size(), it->dist); f_map.erase (s1); break;
        case S_2:
            distplot (f_map.size(), it->dist); f_map.erase(s2);  break;
        }
    }
}

/* ---------------- remove_seeds -----------------------------
 */
static void
remove_seeds (map<string, fseq_prop> &f_map, vector<string> &v_cmt)
{
    vector<string>::const_iterator it = v_cmt.begin();
    for (; it != v_cmt.end(); it++)
        if (it->find (SEED_STR) != string::npos)
            f_map.erase (*it);
}

/* ---------------- squash  ----------------------------------
 * Given a string, keep only the positions which are set
 * true in the vector.
 */
static void
squash (string &s, const vector<bool> &v_used) {
    string new_s;
    new_s.reserve(s.size());
    vector<bool>::const_iterator v_it = v_used.begin();
    for( std::string::iterator s_it=s.begin(); s_it!=s.end(); ++s_it, ++v_it)
        if (*v_it)
            new_s += *s_it;
    s = new_s;
    s.shrink_to_fit();
}

/* ---------------- write_kept_seq ---------------------------
 * This is the final writing of sequences that we want to keep.
 * filter_col means remove gaps that are present in every sequence.
 * r_gaps_flag means remove all gaps.
 */
static int
write_kept_seq (const char *in_fname, const char *out_fname,
                map<string, fseq_prop> &f_map, const vector<bool> &v_used,
                const bool r_gaps_flag)
{
    ifstream in_file (in_fname);
    const char *o_fail_r = "open fail (reading) on ";
    const char *o_fail_w = "open fail for writing on ";
    bool filter_col;
    if (v_used.size() != 0)
        filter_col = true;
    else
        filter_col = false;

    fseq fs;
    if (! in_file)
        return(bust(__func__, o_fail_r, in_fname, 0));

    ofstream out_file (out_fname);
    if ( ! out_file)
        return (bust(__func__, o_fail_w, out_fname, ": ",  strerror(errno), 0));
    if (r_gaps_flag && filter_col)
        return (bust (__func__, "programming bug. Both rgaps and filter_col set", 0));
    
    while (fs.fill (in_file, 0)) {
        const map<string, fseq_prop>::const_iterator missing = f_map.end();
        const map<string, fseq_prop>::const_iterator f1 = f_map.find(fs.get_cmmt());
        if (f1 != missing) {
            out_file << fs.get_cmmt() << '\n'; /* Write comment verbatim */
            size_t done = 0, to_go;   /* but the sequence could have long */
            if (r_gaps_flag)
                fs.clean(false, true); /* Remove gaps and white spaces */
            string s = fs.get_seq();  /* lines that should be split into pieces. */
            if (filter_col)           /* Remove columns that were not used */
                squash (s, v_used);
            else if (r_gaps_flag) {

            }
            to_go = s.length();
            while (to_go) {
                size_t this_line = SEQ_LINE_LEN;
                if (SEQ_LINE_LEN > to_go)
                    this_line = to_go;
                out_file << s.substr (done, this_line)<< '\n';
                done += this_line;
                to_go -= this_line;
            }
            f_map.erase (fs.get_cmmt()); /* stop duplicates being written again */
        }
    }

    in_file.close();
    out_file.close();
    return (EXIT_SUCCESS);
}

/* ---------------- find_used_columns ------------------------
 * We have an alignment, but not all the columns are used.
 * Visit every sequence in the alignment
 * Visit every site in the sequence and mark the corresponding
 * position as true if it is not a gap.
 */
static int
find_used_columns (const char *in_fname,
                   const map<string, fseq_prop> &f_map, vector<bool> &v_used,
                   const short unsigned verbosity)
{
    unsigned nf_in = 0;
    ifstream in_file (in_fname);
    fseq fs;
    if (! in_file)
        return (bust(__func__, "open fail reading from ", in_fname, 0));

    const map<string, fseq_prop>::const_iterator missing = f_map.end();
    while (fs.fill (in_file, 0)) {
        nf_in++;
        const map<string, fseq_prop>::const_iterator f = f_map.find(fs.get_cmmt());
        if (f != missing) {
            const string s = fs.get_seq(); /* need this temporary, otherwise memory error */
            string::const_iterator s_it = s.begin();
            vector<bool>::iterator v_it = v_used.begin();
            for (unsigned short n = 0 ;s_it != s.end(); s_it++, v_it++, n++)
                if (*s_it != GAPCHAR)
                    if (! *v_it)
                        *v_it = true;
        }
    }
    in_file.close();
    if (verbosity > 0) {
        unsigned n = 0;
        vector<bool>::iterator v_it = v_used.begin();
        for (; v_it != v_used.end(); v_it++)
            if (*v_it)
                n++;
        cout << __func__<< ": "<< nf_in << " sequences read. Of " <<
            v_used.size() << " sites, "<< n<< " will be kept.\n";
    }
    return EXIT_SUCCESS;
}

/* ---------------- main  ------------------------------------ */
int
main (int argc, char *argv[])
{
    int c;
    short unsigned verbosity = 0;
    bool seedflag    = false;
    bool filter_col  = true;
    bool eflag       = false;
    bool r_gaps_flag = false;
    bool ignore_len_check = false;
    string choice_name, seed_str;
    unsigned long seed;
    default_random_engine r_engine{};
    const char *progname = argv[0];
    const char *sacred_fname = nullptr;
    const char *plot_fname = nullptr;

    while ((c = getopt(argc, argv, "a:c:e:fgip:sv")) != -1) {
        switch (c) {
        case 'a':
            sacred_fname = optarg;                                     break;
        case 'c':
            choice_name += optarg;                                     break;
        case 'e':
            seed_str = optarg;                                         break;
        case 'f':
            filter_col = false;                                        break;
        case 'g':
            r_gaps_flag = true;                                        break;
        case 'i':
            ignore_len_check = true;                                   break;
        case 'p':
            plot_fname = optarg;                                       break;
        case 's':
            seedflag = true;                                           break;
        case 'v':
            verbosity++;                                               break;
        case ':':
            cerr << argv[0] << "Missing opt argument\n"; eflag = true; break;
        case '?':
            cerr << argv[0] << "Unknown option\n";       eflag = true; break;
        }
    }

    if (r_gaps_flag && filter_col) // column filtering has no effect if we are
        filter_col = false;        // removing gaps anyway

    if (filter_col && ignore_len_check) {
        cerr << "Both column filtering (-f) and ignore length check (-i)"
             << " were turned on.\nYou cannot filter columns unless all sequences\n"
             << "are the same length\n";
        eflag = true;
    }

    if (eflag)
        return (usage(progname, ""));

    if ((argc - optind) < 4)
        return (usage (progname, " too few arguments"));
    const char *in_fname           = argv[optind++];
    const char *dist_fname         = argv[optind++];
    const char *out_fname          = argv[optind++];
    const char *to_keep_str        = argv[optind++];

    unsigned long n_to_keep;
    if (plot_fname)
        try {
            distplot_setup (plot_fname);
        } catch (const std::exception& e) {
            return (bust(progname, "opening ", plot_fname, " for writing:", e.what(), 0));
        }
            
    try {
        n_to_keep = stoul (to_keep_str);
        if (seed_str.length())
            seed = stoul (seed_str);
        else
            seed = DFLT_SEED;
        r_engine.seed( seed);
    } catch (const std::invalid_argument& ia) {
        return(bust(progname, "invalid argument for num seqs: \"", to_keep_str, "\"", ia.what(), 0));
    }
    cout << progname << ": using " << in_fname << " as multiple seq alignment.\nDistance matrix from "
         << dist_fname << "\nWriting to " << out_fname
         << "\nKeeping " << n_to_keep << " of the sequences\n";
    if (plot_fname)
        cout << "Writing a plot file to "<< plot_fname << "\n";
    struct seq_props s_props;
    int gsl_ret;
    thread gsl_thr (get_seq_list, ref(s_props), in_fname, ignore_len_check, &gsl_ret);

    vector<string> v_sacred;
    int sacred_ret = EXIT_SUCCESS;
    thread sac_thr;
    if (sacred_fname)
        sac_thr = thread (get_sacred, sacred_fname, ref(v_sacred), &sacred_ret);

    decider_f *choice = always_first;

    if (choice_name.size()) {
        if ((choice = set_up_choice(choice_name)) == nullptr) {
            gsl_thr.join();
            sac_thr.join();
            return EXIT_FAILURE;
        }
    }

    vector<dist_entry> v_dist; /* Big vector with sorted distance entries */
    vector<string> v_cmt;

    if ( read_distmat (dist_fname, v_dist, v_cmt) == EXIT_FAILURE) {
        cerr << "Waiting on some threads to finish\n";
        return (EXIT_FAILURE);
        gsl_thr.join(); sac_thr.join();
        return (bust(progname, "error reading distance matrix", 0));
    }
    if (verbosity > 0)
        cout << "Finished reading distance matrix\n";
    gsl_thr.join();
    if (gsl_ret != EXIT_SUCCESS) {
        if (sacred_fname)
            sac_thr.join();
        return (bust(progname, "error in get_seq_list", 0));
    }
    if (verbosity > 1)
        cout << "get_seq_list thread finished\n";
    if (check_lists (s_props.f_map, v_cmt) == EXIT_FAILURE) {
        const char *o = "\", original sequences from \"";
        if (sacred_fname)
            sac_thr.join();
        return(bust(progname, "distmat file: \"", dist_fname, o, in_fname, 0));
    }
    if (seedflag)
        remove_seeds (s_props.f_map, v_cmt);
    if (sacred_fname) {
        sac_thr.join();
        if (sacred_ret != EXIT_SUCCESS)
            return (bust(progname, "error reading sacred file from", sacred_fname, 0));
        if (mark_sacred (s_props.f_map, v_sacred) != EXIT_SUCCESS)
            return EXIT_FAILURE;
    }

    remove_seq (s_props.f_map, v_cmt, v_dist, n_to_keep, choice, r_engine);
    distplot_close();
    vector<bool> v_used;
    if (filter_col) {
        v_used.assign (s_props.len, false);
        find_used_columns (in_fname, s_props.f_map, v_used, verbosity);
    } else {
        v_used.resize(0);
    }

    if (write_kept_seq (in_fname, out_fname, s_props.f_map, v_used, r_gaps_flag) != EXIT_SUCCESS)
        return EXIT_FAILURE;

    return EXIT_SUCCESS;
}
